pallets/settlement/src/lib.rs

```diff
@@ -1340,50 +1394,105 @@ impl<T: Trait> Module<T> {
-                                            );
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
-                                                &leg.to.did,
-                                                &leg.to_account_id,
-                                                new_receiver_balance,
-                                            );
+                                        // Get sender's current balance.
+                                        if let Ok((from_mercat, _)) = Self::get_account_and_balance(
+                                            &leg.from.did,
+                                            &leg.from_account_id,
+                                        ) {
+                                            // Verify the proofs.
+                                            let mut rng = rng::Rng::default();
+                                            let result = TransactionValidator {}
+                                                .verify_transaction(
+                                                    &decoded_justified_tx,
+                                                    &from_mercat,
+                                                    &from_mercat_pending_state,
+                                                    &to_mercat,
+                                                    &to_mercat_balance, // If the only reason we need this is updating balance, we should remove it.
+                                                    &[],
+                                                    &mut rng,
+                                                );
+
+                                            if let Ok((_, _)) = result {
+                                                let result = <ConfidentialAsset<T>>::mercat_account_withdraw_amount(
+                                                    &leg.from.did,
+                                                    &leg.from_account_id,
+                                                    decoded_justified_tx
+                                                        .finalized_data
+                                                        .init_data
+                                                        .memo
+                                                        .enc_amount_using_sndr,
+                                                );
+
+                                                // This could be a problem, if we fail to update the storage, but other storage is updated.
+                                                if result.is_err() {
+                                                    return Err((
+                                                        leg_id,
+                                                        FailureReason::GetMercatAccount,
+                                                    ));
+                                                }
```

Comment:
```suggestion
                                                <ConfidentialAsset<T>>::mercat_account_withdraw_amount(
                                                    &leg.from.did,
                                                    &leg.from_account_id,
                                                    decoded_justified_tx
                                                        .finalized_data
                                                        .init_data
                                                        .memo
                                                        .enc_amount_using_sndr,
                                                ).map_err(|_| (leg_id, FailureReason::GetMercatAccount))?;
```
(and in other places)